### PR TITLE
feat: udf reduce

### PR DESCRIPTION
**Pull request by chromevoid**
udf reduce

# Test
## Client
Use numaflow-go sdk client code as the client
```go
package server

import (
	"context"
	"fmt"
	"testing"
	"time"

	fucntionpb "github.com/numaproj/numaflow-go/pkg/apis/proto/function/v1"
	"github.com/numaproj/numaflow-go/pkg/function"
	"github.com/numaproj/numaflow-go/pkg/function/client"
	"github.com/stretchr/testify/assert"
	grpcmd "google.golang.org/grpc/metadata"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func Test_server_reduce(t *testing.T) {
	var ctx = context.Background()
	c, err := client.New(client.WithSockAddr("/tmp/numaflow-test.sock"))
	assert.NoError(t, err)
	defer func() {
		err = c.CloseConn(ctx)
		assert.NoError(t, err)
	}()
	testDatum := &fucntionpb.Datum{
		Key:       "test_id",
		Value:     []byte(`message_success`),
		EventTime: &fucntionpb.EventTime{EventTime: timestamppb.New(time.Unix(1661169600, 0))},
		Watermark: &fucntionpb.Watermark{Watermark: timestamppb.New(time.Time{})},
	}
	ctx = grpcmd.NewOutgoingContext(ctx, grpcmd.New(map[string]string{function.DatumKey: "client", function.WinEndTime: "1662998460", function.WinStartTime: "1662998400"}))
	reduceCh := make(chan *fucntionpb.Datum, 10)
	for i := 0; i < 10; i++ {
		reduceCh <- testDatum
	}
	close(reduceCh)
	response, err := c.ReduceFn(ctx, reduceCh)
	fmt.Println(response, err)
}
```
## Server
Use numaflow-python sdk function example counter code as the server
```python
from typing import Iterator
from pynumaflow.function import Messages, Message, Datum, Metadata, UserDefinedFunctionServicer


def reduce_handler(key: str, datums: Iterator[Datum], md: Metadata) -> Messages:
    interval_window = md.interval_window
    counter = 0
    for datum in datums:
        print("User reduce got", key, datum.value)
        counter = counter + 1
    msg = "counter:%s interval_window_start:%s interval_window_end:%s" % (
        counter,
        interval_window.start,
        interval_window.end,
    )
    messages = Messages()
    messages.append(Message.to_vtx(key, str.encode(msg)))
    return messages


if __name__ == "__main__":
    grpc_server = UserDefinedFunctionServicer(reduce_handler=reduce_handler, sock_path="/tmp/numaflow-test.sock")
    grpc_server.start()
```
# Result
client test output
```sh
=== RUN   Test_server_reduce
[key:"client" value:"counter:10 interval_window_start:2022-09-12 16:00:00+00:00 interval_window_end:2022-09-12 16:01:00+00:00"] <nil>
--- PASS: Test_server_reduce (0.00s)
PASS
```
server reduce udf output
```sh
GRPC Server listening on: unix:///tmp/numaflow-test.sock with max threads: 64
User reduce got client b'message_success'
User reduce got client b'message_success'
User reduce got client b'message_success'
User reduce got client b'message_success'
User reduce got client b'message_success'
User reduce got client b'message_success'
User reduce got client b'message_success'
User reduce got client b'message_success'
User reduce got client b'message_success'
User reduce got client b'message_success'
```

# TODO
After this PR is merged, build the counter example image using the new release and upload.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,27 +6,38 @@ and [UDSinks](https://numaflow.numaproj.io/user-guide/sinks/user-defined-sinks/)
 ## Implement a User Defined Function (UDF)
```
Comment (vigith) [inline README.md:7]:
```suggestion
### Map
```

File: README.md [hunk 2]
```diff
@@ -6,27 +6,38 @@ and [UDSinks](https://numaflow.numaproj.io/user-guide/sinks/user-defined-sinks/)
...
-    Simple UDF that relays an incoming message.
-    """
+def map_handler(key: str, datum: Datum) -> Messages:
     val = datum.value
     _ = datum.event_time
     _ = datum.watermark
-    messages = Messages(Message(key=key, value=val))
+    messages = Messages(Message.to_vtx(key, val))
+    return messages
+
```
Comment (vigith) [inline README.md:18]:
```suggestion
```

### Reduce
```python
```

File: examples/function/counter/Makefile [hunk 3]
```diff
@@ -0,0 +1,3 @@
+.PHONY: image
+image:
+	docker build -t "quay.io/numaio/flatmap-example:python" .
```
Comment (vigith) [inline examples/function/counter/Makefile:3]: is the path right?

File: examples/function/counter/README.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Flatmap Example
+
+## Build image and push
+
+```shell
+make image
+# Privilege needed.
+docker push quay.io/numaio/flatmap-example:python
```
Comment (vigith) [inline examples/function/counter/README.md:8]: shouldn't we do `$language/...` in the path?

File: examples/function/counter/example.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+from typing import Iterator
```
Comment (vigith) [inline examples/function/counter/example.py:1]: separate `map` and `reduce` as two examples?

Comment (chromevoid) [reply examples/function/counter/example.py:1]: no, this map is not required.. committed it accidentally 

deleted

File: pynumaflow/function/server.py [hunk 6]
```diff
@@ -11,55 +12,90 @@
...
 _LOGGER = logging.getLogger(__name__)
 
 UDFMapCallable = Callable[[str, Datum], Messages]
+UDFReduceCallable = Callable[[str, Iterator[Datum], Metadata], Messages]
 _PROCESS_COUNT = multiprocessing.cpu_count()
 MAX_THREADS = int(os.getenv("MAX_THREADS", 0)) or (_PROCESS_COUNT * 4)
 
 
+def default_map_handler(_: str, __: Datum) -> Messages:
+    return
```
Comment (chromevoid) [inline pynumaflow/function/server.py:37]: I want to make input arguments map_handler and reduce_handler to be optional.
That's why I added these two `default_handler` functions.

Need suggestions.. whether there's a more standard way to solve this?

Comment (vigith) [reply pynumaflow/function/server.py:37]: I don't think you can do that at compile time. Python's type checking system is not that mature. @ab93 ?

Comment (chromevoid) [reply pynumaflow/function/server.py:37]: > I don't think you can do that at compile time. Python's type checking system is not that mature. @ab93 ?

Tested and passed. Does it mean compile succeeded? https://github.com/numaproj/numaflow-python/pull/40#issue-1500993188

Comment (chromevoid) [reply pynumaflow/function/server.py:37]: if I call both reduce function and map function in the above test
```go
response, err := c.ReduceFn(ctx, reduceCh)
fmt.Println(response, err)
response, err = c.MapFn(ctx, testDatum)
fmt.Println(response, err)
```
got
```sh
[key:"client" value:"counter:10 interval_window_start:2022-09-12 16:00:00+00:00 interval_window_end:2022-09-12 16:01:00+00:00"] <nil>
[] failed to execute c.grpcClt.MapFn(): rpc error: code = Unknown desc = Exception calling application: 'NoneType' object has no attribute 'items'
```

Comment (vigith) [reply pynumaflow/function/server.py:37]: so, if the user passes the wrong handler, we can catch that only a runtime.

Comment (chromevoid) [reply pynumaflow/function/server.py:37]: Wrong handler?
You mean user create a reduce udf but use it as a map udf? Then yes the error can only be triggered when the apply udf is called

Comment (vigith) [reply pynumaflow/function/server.py:37]: yes

File: pynumaflow/tests/function/test_server.py [hunk 7]
```diff
@@ -175,17 +254,51 @@ def test_reduce_fn(self):
...
+            response.elements[0].value,
+        )
+
+    def test_invalid_input(self):
+        try:
+            _ = UserDefinedFunctionServicer()
+        except Exception as err:
+            self.assertEqual(
+                "Require a valid map handler and/or a valid reduce handler.", err.__str__()
+            )
```
Comment (ab93) [inline pynumaflow/tests/function/test_server.py:284]:
```suggestion
        with self.assertRaises(ValueError):
            UserDefinedFunctionServicer()
```

File: examples/function/counter/example.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    for _ in datums:
+        counter = counter + 1
+    msg = "counter:%s interval_window_start:%s interval_window_end:%s" % (
+        counter,
+        interval_window.start,
+        interval_window.end,
+    )
+    messages = Messages()
+    messages.append(Message.to_vtx(key, str.encode(msg)))
+    return messages
```
Comment (ab93) [inline examples/function/counter/example.py:17]: let's make this simpler as discussed

**Review by ab93 — CHANGES_REQUESTED** [review]

File: examples/function/counter/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,43 @@
+FROM python:3.9.12-slim
```
Comment (ab93) [inline examples/function/counter/Dockerfile:1]: Can you copy the optimized dockerfile from numalogic? It is under examples.

**Review by ab93 — COMMENTED** [review]
Minor comment

**Review by ab93 — APPROVED** [review]
Looks good

**Pull request merged**